crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2304,34 +2326,31 @@ impl<'db> FunctionType<'db> {
     ///
     /// Were this not a salsa query, then the calling query
     /// would depend on the function's AST and rerun for every change in that file.
-    #[salsa::tracked]
-    pub fn return_ty(self, db: &'db dyn Db) -> Type<'db> {
+    #[salsa::tracked(return_ref)]
+    pub fn external_signature(self, db: &'db dyn Db) -> Signature<'db> {
+        let function_stmt_node = self.body_scope(db).node(db).expect_function();
+        let internal_signature = self.internal_signature(db);
+        if function_stmt_node.decorator_list.is_empty() {
+            return internal_signature;
+        }
+        // TODO process the effect of decorators on the signature
+        Signature::todo()
+    }
+
+    /// Typed internally-visible signature for this function.
+    ///
+    /// This represents the annotations on the function itself, unmodified by decorators.
+    ///
+    /// These are the argument and return types that should be used for type checking the body of
+    /// the function.
+    ///
+    /// Don't call this when checking any other file; only when type-checking the function body
+    /// scope.
```

Comment:
Hmm, this is an interesting thought; I'll consider what that could look like.

Reply:
It's hard to say where else it can/should go until we have the users for it (e.g. inferring types of function parameters, verifying returned types.) So I'm going to leave it here for now.